marlowe/specification/ReadMe.md

```diff
@@ -0,0 +1,346 @@
+```haskell
+marloweState marloweData ≡ inState
+```
+
+
+#### *Constraint 8.* Input contract
+
+The input contract of the transaction is used in the semantics computation.
+```haskell
+marloweContract marloweData ≡ inContract
+```
+
+
+#### *Constraint 9.* Marlowe parameters
+
+The Marlowe parameters are not changed by the transaction.
+```haskell
+Just $ marloweParams marloweData ≡ fmap marloweParams . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 10.* Output state
+
+The output state of the transaction matches the result of the semantics computation.
+```haskell
+Just $ txOutState transactionOutput ≡ fmap marloweState . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 11.* Output contract
+
+The output contract of the transaction matches the result of the semantics computation.
+```haskell
+Just $ txOutContract transactionOutput ≡ fmap marloweContract . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 12.* Merkleized continuations
+
+All of the merkleized contracts referenced in the input are provided to the script.
+```haskell
+all demerkleizes transactionInput ≡ True
+  where demerkleizes (MerkelizedTxInput _ hash) = isJust (fromBuiltinData =<< lookup hash (txInfoData $ scriptContextTxInfo scriptContext) :: Maybe Contract)
+        demerkleizes (Input _) = True
+```
+
+
+#### *Constraint 13.* Positive balances
+
+All accounts in the state have positive balances.
+```haskell
+all ((> 0) . snd) (accounts $ marloweState marloweData) ≡ True
+```
+
+
+#### *Constraint 14.* Inputs authorized
+
+The required signature or at-least-one role token is present in the transaction for each deposit or choice.
+```haskell
+all authorized (fmap getInputContent transactionInput) ≡ True
+  where getInputContent (Input inputContent) = inputContent
+        getInputContent (MerkleizedTxInput inputContent _) = inputContent
+        authorized (IDeposit _ (PK hash) _ _) = authorizedHash hash
+        authorized (IDeposit _ (Role role) _) = authorizedRole role
+        authorized (IChoice (ChoiceId _ (PK hash)) _) = authorizedHash hash
+        authorized (IChoice (ChoiceId _ (Role role)) _) = authorizedRole role
+        authorized INotify = True
+        authorizedHash hash = elem hash (txInfoSignatories $ scriptContextTxInfo scriptContext)
+        authorizedRole role = Value.leq (Value.singleton (rolesCurrency $ marloweparams marloweData) role 1) spent
+        spent = foldMap (txOutValue . txInInfoResolved) (txInfoInputs $ scriptContextTxInfo scriptContext)
+        -- FIXME: We should write out `Value.leq` ourselves, so the specification does not depend upon Plutus's convenience functions.
+```
+
+
+#### *Constraint 15.* Sufficient payment
+
+Sufficient funds are paid to the public keys or roles.
+```haskell
+all sufficient (txOutPayments transactionOutput)
+  where sufficient (Payment _ (Party (PK hash)) value) = Value.leq value (foldMap (paidToHash hash) (txInfoOutputs $ scriptContextTxInfo scriptContext))
```

Comment:
Now I get why it is not obvious how `leq` behaves, it is the Map of Maps artefact which is tricky to handle because it is not just a number

Reply:
Also it seems that `foldMap` is working on the assumption that `Value` is a monoid that has addition as operation? Probably would be good to be explicit. Definitely this specification is helping me understand things better, nice :)

Reply:
Is this the tightest possible spec? Under what circumstances is the relation > rather than =? Would be worth including rationale for this

Reply:
Thanks for these comments. I'll make the following changes:
1. Specify the monoid laws that `Value` must conform to.
2. Specify the behavior required of `leq`. (I believe that `Value` comparison was the cause of the MinSwap vulnerability, so we need to be really clear about comparisons.)
3. Add the rationale for `>` instead of `==`. Briefly, `>` is required so that transactions conform to the min-ada requirement. For example, it is impossible to send less than one Ada or to send a token w/o any Ada.

Reply:
If we want to handle multiple Marlowe validators with distinct role currencies we can mark PK payouts using `datum` containing `roleCurrency` value so we can prevent double spending I think.

Reply:
@paluh, yes, I'll leave a `FIXME` in the document and re-sync it with the validator once that is frozen.

Reply:
@simonjohnthompson, I will tighten up the specification slightly by defining `leq` to require `==` for Ada but `<=` for non-Ada tokens. We cannot make this full equality because of the min-UTxO ledger constraint.

Reply:
fixed in 8be04323e4ec24c4e923ed214094e85c916ab8e0